### DISCUSSION TITLE
Improvements to the typing infrastructure

### DISCUSSION THREAD

**Discussion by EmilyBourne** [pyccel/pyccel#1722] [Ideas]
## Current situation and limitations

Typed objects in Pyccel are currently represented in a `TypedAstNode`. Each `TypedAstNode` has the following properties:
- datatype
- precision
- rank
- shape
- order
- class type

As we add support for various containers the limitations of this approach are becoming evident. Certain objects cannot be represented with this information. For an example, see #1721 : 
```python
a = [np.array((2,3), order='F')]
```
As lists use C ordering, the ordering of the final object cannot be expressed as simply `C` or `F`. Another example is a `list[set[int]]`. The datatype is `int` but currently the class type does not hold a reference to the associated datatype.

This could be added relatively easily. It has the advantage of making the class type mimic Python's types more closely. However it seems strange to mimics Python's types only partially. For example `list[int]` and `list[int64]` would have the same class type.

Another place which causes issues is the rank of lists/sets/tuples. Logically these objects are 1D and may contain ND objects, however as the code is currently written a list of arrays must be described as a 2D object to ensure that no information is lost.

A final example is that of inhomogeneous tuples. The type `tuple[int, float[:]]` cannot be described. Currently the code works round this by saving the rank to 1 and the class type to `tuple[int, np.ndarray]`. As the `tuple` is immutable the elements can be created on initialisation with the correct properties avoiding the problems that arise from the incomplete description. It is then necessary for the code to always collect the element of the tuple before deducing any information related to it.

## Suggestion

I suggest removing this information from the AST definition. Instead I think we should have more complete types in datatypes.py. I suggest that we break the datatypes into 3:
- `PrimitiveType`
- `FixedSizeType`
- `ContainerType`

The `PrimitiveType` would be one of:
- bool
- int
- float
- complex
- char

A `FixedSizeType` would contain:
- name
- Datatype
- precision

A `ContainerType` would contain:
- name
- `element_type : SignedType | ContainerType`
- rank
- order

A `TypedAstNode` would then simply contain a type (`SignedType | ContainerType`) and a shape. The shape should have the same length as the rank of the outermost container. The shape of elements would simply use `PyccelArrayShapeElement`.

This solution stays very close to the current implementation which should facilitate its implementation but it gives additional information as it is then possible to describe `list[set[int]]`, `list[np.ndarray[int](order=F)]`, and `tuple[int, np.ndarray[float]]`

**Comment by EmilyBourne** [comment]
cc: @yguclu @Farouk-Echaref

**Reply by yguclu** [reply]
Sounds very reasonable to me.

**Reply by yguclu** [reply]
Why the name choice of `SignedType`?

**Reply by EmilyBourne** [reply]
Honestly I was struggling to find good names for these concepts. I googled and tried to find existing names, but you are right this is a terrible name as it is not about signed/unsigned. Do you have a suggestion for something better?

**Reply by yguclu** [reply]
What does that class represent in your mind?

**Reply by yguclu** [reply]
Scalar variables maybe?

**Reply by EmilyBourne** [reply]
But these are not variables they are types. `BasicType`?

**Reply by EmilyBourne** [reply]
> What does that class represent in your mind?

It represents the datatype in a low-level language. Things like:
- int32
- int64 
- float
- double

Maybe it should just be `DataType` but then what do you call an integer of unknown precision?

**Reply by yguclu** [reply]
I see. `SignedType` is a built-in scalar datatype which you can actual represent in memory. The integer of unknown precision cannot be.

**Reply by EmilyBourne** [reply]
`PrimitiveType` and `FixedSizeType` ?

**Comment by EmilyBourne** [comment]
Additional example:
The annotation `tuple[int32, int64]` cannot be described with the current typing system

**Comment by said-hadjout** [comment]
we also need to add SIMD vector datatypes which are described here https://docs.modular.com/mojo/stdlib/builtin/simd.html

**Reply by EmilyBourne** [reply]
I am not sure that we **need** to support this, but it could be a nice thing to have and improvements to the typing system should make this easier to support.

What C/Fortran data structure does this correspond to? There is a discussion here : https://www.infoworld.com/article/3175127/simd-intrinsics-aren-t-so-scary-but-should-we-use-them.html
about SIMD in C. There are 4 suggestions. The first 3 (Use libraries e.g. MKL, Use compiler flags, Use OpenMP) are already supported by Pyccel. The fourth does not seem to refer to types but function calls, and it is not recommended. Are you thinking of something else?

**Reply by said-hadjout** [reply]
Im thinking about using OpenMP , in the past each compiler had its own directives , for more details see this comment https://fortran-lang.discourse.group/t/fortran-syntax-for-pragmas/6532/8 .
we should be able to add the relevant flags automatically either by using OpenMP directives or if we want , we can use compiler specific directives

**Reply by said-hadjout** [reply]
we should be able to imitate the behavior explained here by mojo https://www.youtube.com/watch?v=3OWkXNdkx8E&t=487s,
and I think we have all the necessary ingredients for it

**Reply by EmilyBourne** [reply]
Ok. In that case I think you should open a separate discussion as I'm not sure how OpenMP pragmas relate to the way of internally keeping track of datatypes and this discussion is likely to get lost. We currently have support for the `!$omp simd` pragma recommended by Ivan.

There are multiple places where it could be interesting to add annotations automatically. This could be for general parallelism (e.g. OpenMP) or for SIMD instructions or for the pure decorator, etc. However we are not currently focusing on optimisations. Whenever we have mentioned this kind of improvement recently we have decided that we prefer to consolidate our basic support by finishing support for:
- lists
- sets
- dicts
- strings

Once we have this then we can start to look at optimisations like automatic OpenMP insertion

**Reply by EmilyBourne** [reply]
The tensor discussion in the video may be relevant for @jalalium

**Reply by said-hadjout** [reply]
Sure, I just wanted to make sure that the new datatypes infrastructure can handle these kind of types

**Reply by EmilyBourne** [reply]
My aim is to make this sufficiently general that extensions (e.g. cuda) can be handled through this. I imagine that a SIMD type would either be described by a `ContainerType` which takes 1 argument. So you would have something like `SIMD[float64]` or they would be described by a `FixedSizeType` and would work alongside NumPy types as similar but distinct types `SIMD_float64() != float64()`

**Comment by said-hadjout** [comment]
It would also be nice if the user can define a type of variable rank nd_arrays that is a union of nd_arrays of multiple ranks
in the current syntax I thought about something like `int[1:3]` which means `int[:]|int[:,:]|int[:,:,:]`

**Reply by EmilyBourne** [reply]
This sounds like an interesting idea. I am not sure about the syntax, this should be discussed. Please create a dedicated discussion. This is unrelated to the subject discussed here and could already be implemented with our current datatype organisation
